rescan_test.go

```diff
@@ -0,0 +1,680 @@
+// notifications should be delivered from this height. When providing a
+// bestHeight of 0, a backlog will not be delivered.
+func (c *mockChainSource) Subscribe(
+	bestHeight uint32) (*blockntfns.Subscription, error) {
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.subscribeOverride != nil {
+		return c.subscribeOverride(bestHeight)
+	}
+
+	for i := bestHeight + 1; i <= uint32(c.bestBlock.Height); i++ {
+		hash := c.blockHashesByHeight[i]
+		header := c.blockHeaders[*hash]
+		c.ntfnChan <- blockntfns.NewBlockConnected(*header, i)
+	}
+
+	return &blockntfns.Subscription{
+		Notifications: c.ntfnChan,
+		Cancel:        func() {},
+	}, nil
+}
+
+// rescanTestContext serves as a harness to aid testing the Rescan struct.
+type rescanTestContext struct {
+	t                  *testing.T
+	chain              *mockChainSource
+	blocksConnected    chan waddrmgr.BlockStamp
+	blocksDisconnected chan waddrmgr.BlockStamp
+	rescan             *Rescan
+	errChan            <-chan error
+	quit               chan struct{}
+}
+
+// newRescanTestContext creates a new test harness for the Rescan struct backed
+// by a chain of numBlocks.
+func newRescanTestContext(t *testing.T, numBlocks int,
+	options ...RescanOption) *rescanTestContext {
+
+	blocksConnected := make(chan waddrmgr.BlockStamp)
+	blocksDisconnected := make(chan waddrmgr.BlockStamp)
+	ntfnHandlers := rpcclient.NotificationHandlers{
+		OnFilteredBlockConnected: func(height int32,
+			header *wire.BlockHeader, _ []*btcutil.Tx) {
+
+			blocksConnected <- waddrmgr.BlockStamp{
+				Hash:      header.BlockHash(),
+				Height:    height,
+				Timestamp: header.Timestamp,
+			}
+		},
+		OnFilteredBlockDisconnected: func(height int32,
+			header *wire.BlockHeader) {
+
+			blocksConnected <- waddrmgr.BlockStamp{
+				Hash:      header.BlockHash(),
+				Height:    height,
+				Timestamp: header.Timestamp,
+			}
+		},
+	}
+	quit := make(chan struct{})
+
+	chain := newMockChainSource(numBlocks)
+	rescanOptions := []RescanOption{
+		NotificationHandlers(ntfnHandlers), QuitChan(quit),
+	}
+	rescanOptions = append(rescanOptions, options...)
+	rescan := NewRescan(chain, rescanOptions...)
+
+	return &rescanTestContext{
+		t:                  t,
+		chain:              chain,
+		blocksConnected:    blocksConnected,
+		blocksDisconnected: blocksDisconnected,
+		rescan:             rescan,
+		quit:               quit,
+	}
+}
+
+// start starts the backing rescan.
+func (ctx *rescanTestContext) start() {
+	ctx.errChan = ctx.rescan.Start()
+}
+
+// stop stops the rescan test harness and ensures it exits cleanly.
+func (ctx *rescanTestContext) stop() {
+	close(ctx.quit)
+
+	select {
+	case err := <-ctx.errChan:
+		if err != ErrRescanExit {
+			ctx.t.Fatalf("expected ErrRescanExit upon shutdown, "+
+				"got %v", err)
+		}
+
+	case <-time.After(time.Second):
+		// Print a goroutine dump before exiting.
+		pprof.Lookup("goroutine").WriteTo(os.Stdout, 1)
+		ctx.t.Fatal("rescan did not exit cleanly")
+	}
+}
```

Comment:
Check here that no unexpected events have been posted into the channels?

Reply:
Done.